### ISSUE TITLE
tf.keras.backend.set_floatx() causing ValueError (dtype conversion error) while computing tf.keras.metrics.*

### ISSUE THREAD

**Issue by Hemal-Mamtora** [tensorflow/tensorflow#36790]
**System information** - Have I written custom code on **Google Colab**: - 
**Code:**
```
tf.keras.backend.set_floatx('float64')

model.compile(optimizer= Adam(learning_rate= 0.001, clipnorm=1.0, clipvalue=0.5),
              loss={
                  'class_output': BinaryCrossentropy(),
                  'decoder_output': BinaryCrossentropy()
              },
              loss_weights=[0.5, 1.0],
               metrics = {
                  'class_output':[tf.metrics.Recall(), tf.metrics.Precision()],
                  'decoder_output':[tf.metrics.Recall(), tf.metrics.Precision()],
              }
             ) 
```
**Error:**
```
/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/ops.py in convert_to_tensor(value, dtype, name, as_ref, preferred_dtype, dtype_hint, ctx, accepted_result_types)
   1288       raise ValueError(
   1289           "Tensor conversion requested dtype %s for Tensor with dtype %s: %r" %
-> 1290           (dtype.name, value.dtype.name, value))
   1291     return value
   1292 

ValueError: Tensor conversion requested dtype float64 for Tensor with dtype float32: <tf.Tensor 'metrics_6/class_output_recall_8/Sum:0' shape=(1,) dtype=float32>
```

OS Platform and Distribution : - 
```
os.uname()
>>> posix.uname_result(sysname='Linux', nodename='ed841897617b', release='4.14.137+', version='#1 SMP Thu Aug 8 02:47:02 PDT 2019', machine='x86_64')
```
TensorFlow installed from : - 
```!pip install tensorflow==2.10```
TensorFlow version : - 
```
tf.__version__
>>> '2.1.0'
```
Python version: - 
```
!python -V
>>> Python 3.6.9
```
~~Bazel version :-  NA~~
~~GCC/Compiler version : - NA~~  
~~CUDA/cuDNN version: -  NA~~

GPU model and memory:
```
from psutil import virtual_memory
mem = virtual_memory()
print(mem.total / 1024**3, 'GB') # total physical memory available
>>>12.717426300048828 GB
````
~~You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`~~

**Describe the current behavior**
When using `tf.keras.backend.set_floatx('float64')` , whole tf should be set to float64, right ?
But the tf.metrics are not getting set as shown in the code above

**Describe the expected behavior**
All of tf including tf.metrics should be calculated on the basis of tf.keras.backend.set_floatx('float64')

**Code to reproduce the issue** 
```
import tensorflow as tf 

tf.keras.backend.set_floatx('float64')

m = tf.keras.metrics.Recall()
m.update_state([0, 1, 1, 1], [1, 0, 1, 1])
print('Final result: ', m.result().numpy())
```

**Other info / logs** ~~Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.~~
StackTrace:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-30-86f79f751766> in <module>()
      6               loss_weights=[0.5, 1.0],
      7                metrics = {
----> 8                   'class_output':[tf.metrics.Recall(), tf.metrics.Precision()],
      9                 #  'decoder_output':[tf.metrics.Recall(), tf.metrics.Precision()],
     10               }

13 frames
/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/training/tracking/base.py in _method_wrapper(self, *args, **kwargs)
    455     self._self_setattr_tracking = False  # pylint: disable=protected-access
    456     try:
--> 457       result = method(self, *args, **kwargs)
    458     finally:
    459       self._self_setattr_tracking = previous_value  # pylint: disable=protected-access

/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/training.py in compile(self, optimizer, loss, metrics, loss_weights, sample_weight_mode, weighted_metrics, target_tensors, distribute, **kwargs)
    437           targets=self._targets,
    438           skip_target_masks=self._prepare_skip_target_masks(),
--> 439           masks=self._prepare_output_masks())
    440 
    441       # Prepare sample weight modes. List with the same length as model outputs.

/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/training.py in _handle_metrics(self, outputs, targets, skip_target_masks, sample_weights, masks, return_weighted_metrics, return_weighted_and_unweighted_metrics)
   2002           metric_results.extend(
   2003               self._handle_per_output_metrics(self._per_output_metrics[i],
-> 2004                                               target, output, output_mask))
   2005         if return_weighted_and_unweighted_metrics or return_weighted_metrics:
   2006           metric_results.extend(

/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/training.py in _handle_per_output_metrics(self, metrics_dict, y_true, y_pred, mask, weights)
   1953       with K.name_scope(metric_name):
   1954         metric_result = training_utils.call_metric_function(
-> 1955             metric_fn, y_true, y_pred, weights=weights, mask=mask)
   1956         metric_results.append(metric_result)
   1957     return metric_results

/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/engine/training_utils.py in call_metric_function(metric_fn, y_true, y_pred, weights, mask)
   1153 
   1154   if y_pred is not None:
-> 1155     return metric_fn(y_true, y_pred, sample_weight=weights)
   1156   # `Mean` metric only takes a single value.
   1157   return metric_fn(y_true, sample_weight=weights)

/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/metrics.py in __call__(self, *args, **kwargs)
    194     from tensorflow.python.keras.distribute import distributed_training_utils  # pylint:disable=g-import-not-at-top
    195     return distributed_training_utils.call_replica_local_fn(
--> 196         replica_local_fn, *args, **kwargs)
    197 
    198   @property

/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/distribute/distributed_training_utils.py in call_replica_local_fn(fn, *args, **kwargs)
   1133     with strategy.scope():
   1134       return strategy.extended.call_for_each_replica(fn, args, kwargs)
-> 1135   return fn(*args, **kwargs)
   1136 
   1137 

/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/metrics.py in replica_local_fn(*args, **kwargs)
    177     def replica_local_fn(*args, **kwargs):
    178       """Updates the state of the metric in a replica-local context."""
--> 179       update_op = self.update_state(*args, **kwargs)  # pylint: disable=not-callable
    180       with ops.control_dependencies([update_op]):
    181         result_t = self.result()  # pylint: disable=not-callable

/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/utils/metrics_utils.py in decorated(metric_obj, *args, **kwargs)
     74 
     75     with tf_utils.graph_context_for_symbolic_tensors(*args, **kwargs):
---> 76       update_op = update_state_fn(*args, **kwargs)
     77     if update_op is not None:  # update_op will be None in eager execution.
     78       metric_obj.add_update(update_op)

/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/metrics.py in update_state(self, y_true, y_pred, sample_weight)
   1340         top_k=self.top_k,
   1341         class_id=self.class_id,
-> 1342         sample_weight=sample_weight)
   1343 
   1344   def result(self):

/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/utils/metrics_utils.py in update_confusion_matrix_variables(variables_to_update, y_true, y_pred, thresholds, top_k, class_id, sample_weight, multi_label, label_weights)
    438       update_ops.append(
    439           weighted_assign_add(label, pred, weights_tiled,
--> 440                               variables_to_update[matrix_cond]))
    441 
    442   return control_flow_ops.group(update_ops)

/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/keras/utils/metrics_utils.py in weighted_assign_add(label, pred, weights, var)
    414     if weights is not None:
    415       label_and_pred *= weights
--> 416     return var.assign_add(math_ops.reduce_sum(label_and_pred, 1))
    417 
    418   loop_vars = {

/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/ops/resource_variable_ops.py in assign_add(self, delta, use_locking, name, read_value)
    783     with _handle_graph(self.handle), self._assign_dependencies():
    784       assign_add_op = gen_resource_variable_ops.assign_add_variable_op(
--> 785           self.handle, ops.convert_to_tensor(delta, dtype=self.dtype),
    786           name=name)
    787     if read_value:

/usr/local/lib/python3.6/dist-packages/tensorflow_core/python/framework/ops.py in convert_to_tensor(value, dtype, name, as_ref, preferred_dtype, dtype_hint, ctx, accepted_result_types)
   1288       raise ValueError(
   1289           "Tensor conversion requested dtype %s for Tensor with dtype %s: %r" %
-> 1290           (dtype.name, value.dtype.name, value))
   1291     return value
   1292 

ValueError: Tensor conversion requested dtype float64 for Tensor with dtype float32: <tf.Tensor 'metrics_6/class_output_recall_8/Sum:0' shape=(1,) dtype=float32>
```

**Comment by Hemal-Mamtora** [comment]
@MarkDaoust 
Tagging you for escalation. Kindly excuse me if it is unprofessional.

**Comment by Saduf2019** [comment]
@Hemal-Mamtora Could you please confirm if the issue faced by you is similar to existing [issue](https://github.com/tensorflow/tensorflow/issues/33365)

**Comment by Hemal-Mamtora** [comment]
Yes, seems like TF 2.0 has issues with float 64, but data comes in float 64,

Till when would this issue he resolved ?

**Comment by MarkDaoust** [comment]
Hi @Hemal-Mamtora , 

I think you're right.

It looks like this is being caused by the mismatch of [metric.py recognizing `floatx`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/metrics.py#L146)  but [metric_utils.py casting directly to float32](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/utils/metrics_utils.py#L427) .

@pavithrasv, what's the right way to fix this?

**Comment by pavithrasv** [comment]
Thank you @MarkDaoust. It should be cast to the predictions' dtype. If anyone would like to work on the fix please feel free to send me a PR.

**Comment by yongtang** [comment]
Added a fix and test case in PR #39134 that will address this issue.
